### PR TITLE
SemiGroup

### LINKED ISSUE #1921: Type Classes: SemiGroup
- [ ] Add associative law.
- [ ] Add `combineN` function to the type class.
- [ ] Add instance for `Option`.
- [ ] Add instance for `List`.
- [ ] Add instance for `Set`.
- [ ] Add instance for `Map`.
- [ ] Add instance for `Validation`.
- [ ] Add instances for tuples up to N = 10.

More information can be found here:

- https://github.com/typelevel/cats/blob/main/kernel/src/main/scala/cats/kernel/Semigroup.scala
- https://typelevel.org/cats/typeclasses/semigroup.html
- https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Semigroup.html
- You may also want to look at `Monoid.flix`

### PR DESCRIPTION
**Pull request by ninaandrup**
Fixes #1921

### CODE REVIEW

File: main/src/library/SemiGroup.flix [hunk 1]
```diff
@@ -31,3 +38,86 @@ instance SemiGroup[Unit] {
 instance SemiGroup[String] {
     pub def combine(x: String, y: String): String = x + y
 }
+
+instance SemiGroup[Option[a]] with SemiGroup[a] {
+  pub def combine(x: Option[a], y: Option[a]): Option[a] = match (x, y) {
+        case (Some(x1), Some(y1)) => Some(SemiGroup.combine(x1, y1))
+        case (a, None) => a
+        case (None, b) => b
```
Comment (magnus-madsen) [inline main/src/library/SemiGroup.flix:46]: I would move the (Some, Some) case last in the pattern match.

File: main/src/library/SemiGroup.flix [hunk 2]
```diff
@@ -31,3 +38,86 @@ instance SemiGroup[Unit] {
...
+instance SemiGroup[List[a]] {
+  pub def combine(x: List[a], y: List[a]): List[a] = x ::: y
+}
+
+instance SemiGroup[Set[a]] with Order[a] {
+  pub def combine(x: Set[a], y: Set[a]): Set[a] = Set.union(x,y)
+}
+
+instance SemiGroup[Map[k,v]] with Order[k], SemiGroup[v] {
+  pub def combine(x: Map[k,v], y: Map[k,v]): Map[k,v] = Map.unionWith((v1, v2) -> SemiGroup.combine(v1, v2), x, y)
```
Comment (magnus-madsen) [inline main/src/library/SemiGroup.flix:59]:
```
Map.unionWith((v1, v2) -> SemiGroup.combine(v1, v2), x, y)
```

should be the same as just

```
Map.unionWith(SemiGroup.combine, x, y)
```

which is more elegant imho :)

File: main/src/library/SemiGroup.flix [hunk 3]
```diff
@@ -31,3 +38,86 @@ instance SemiGroup[Unit] {
...
+}
+
+instance SemiGroup[Validation[t, e]] with SemiGroup[t] {
+  pub def combine(x: Validation[t, e], y: Validation[t, e]): Validation[t, e] = match (x, y) {
+        case (Success(x1), Success(y1)) => Success(SemiGroup.combine(x1, y1))
+        case (Failure(es1), Failure(es2)) => Failure(Nel.append(es1, es2))
+        case (Failure(_), _) => x
+        case (_, Failure(_)) => y
```
Comment (magnus-madsen) [inline main/src/library/SemiGroup.flix:67]: We usually align all the `=>` arrows, i.e. they should line up by indenting the cases.

**Review by magnus-madsen — COMMENTED** [review]
I added some minor comments. In addition I think we should add the associative law.

Other than that, this looks ready to me 👍

**Comment by ninaandrup** [comment]
> I added some minor comments. In addition I think we should add the associative law.
> 
> Other than that, this looks ready to me 👍

**Comment by ninaandrup** [comment]
I was too quick and didn't realise what I thought was "cancel comment" was "close with comment"

**Comment by ninaandrup** [comment]
See #2579

**Pull request closed without merging**